test/x509_time_test.c

```diff
@@ -424,11 +424,121 @@ static int test_days(int n)
     return r;
 }
 
+static const struct {
+    ASN1_TIME asn1;
+    const char *readable;
+} x509_print_tests [] = {
+    {
+        /* Generalized Time */
+        {
+            sizeof("20170731222050Z") - 1,
+            V_ASN1_GENERALIZEDTIME,
+            (unsigned char *)"20170731222050Z",
+            0
+        },
+        "Jul 31 22:20:50 2017 GMT"
+    },
+    {
+        /* Generalized Time, no seconds */
+        {
+            sizeof("201707312220Z") - 1,
+            V_ASN1_GENERALIZEDTIME,
+            (unsigned char *)"201707312220Z",
+            0
+        },
+        "Jul 31 22:20:00 2017 GMT"
+    },
+    {
+        /* Generalized Time, fractional seconds (3 digits) */
+        {
+            sizeof("20170731222050.123Z") - 1,
+            V_ASN1_GENERALIZEDTIME,
+            (unsigned char *)"20170731222050.123Z",
+            0
+        },
+        "Jul 31 22:20:50.123 2017 GMT"
+    },
+    {
+        /* Generalized Time, fractional seconds (1 digit) */
+        {
+            sizeof("20170731222050.1Z") - 1,
+            V_ASN1_GENERALIZEDTIME,
+            (unsigned char *)"20170731222050.1Z",
+            0
+        },
+        "Jul 31 22:20:50.1 2017 GMT"
+    },
+    {
+        /* Generalized Time, fractional seconds (0 digit) */
+        {
+            sizeof("20170731222050.Z") - 1,
+            V_ASN1_GENERALIZEDTIME,
+            (unsigned char *)"20170731222050.Z",
+            0
+        },
+        "Bad time value"
+    },
+    {
+        /* UTC Time */
+        {
+            sizeof("170731222050Z") - 1,
+            V_ASN1_UTCTIME,
+            (unsigned char *)"170731222050Z",
+            0
+        },
+        "Jul 31 22:20:50 2017 GMT"
+    },
+    {
+        /* UTC Time, no seconds */
+        {
+            sizeof("1707312220Z") - 1,
+            V_ASN1_UTCTIME,
+            (unsigned char *)"1707312220Z",
+            0
+        },
+        "Jul 31 22:20:00 2017 GMT"
+    }
+};
+
+static int test_x509_time_print(int idx)
+{
+    BIO *m;
+    int ret = 0, rv;
+    char *pp;
+    const char *readable;
+
+    m = BIO_new(BIO_s_mem());
+    if (m == NULL)
+        goto err;
+
+    rv = ASN1_TIME_print(m, &x509_print_tests[idx].asn1);
+    if (!TEST_int_ge(rv, 0)) {
+        /* this is an BIO_printf error */
+        goto err;
+    }
+    readable = x509_print_tests[idx].readable;
+    if (rv == 0 && !TEST_str_eq(readable, "Bad time value")) {
+        /* only if the test case intends to fail... */
+        goto err;
+    }
+    rv = BIO_get_mem_data(m, &pp);
+    if (!TEST_int_ne(rv, 0)
```

Comment:
I have rolled up all of these. Though from my point of view, sometimes too many 'roll-up' stuffs could also lead to hard-readable code, since the function is called inside another statement such as `if, while...`, this is not so obvious compared to just leave it called alone in one line.